packages/eslint-plugin/src/rules/shorthand-property-sorting/index.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  shorthandBuckets,
+  shorthandFor,
+  kebabCase,
+  type ShorthandProperties,
+} from '@compiled/utils';
+import type { Rule, Scope } from 'eslint';
+import type { ObjectExpression, Property } from 'estree';
+
+import { isStyled, isCss, isCssMap } from '../../index';
+
+const fixProperties = (context: Rule.RuleContext, node: Rule.Node) => {
+  context.report({
+    node: node,
+    messageId: 'shorthand-first',
+    fix: (fixer) => {
+      if (node.type === 'ObjectExpression' && node.properties.length > 0) {
+        // sort the properties by depth
+        const sortedProperties = node.properties.slice().sort((a, b) => {
+          if (
+            a.type === 'Property' &&
+            a.key.type === 'Identifier' &&
+            b.type === 'Property' &&
+            b.key.type === 'Identifier'
+          ) {
+            const propA = kebabCase(a.key.name) as ShorthandProperties;
+            const propB = kebabCase(b.key.name) as ShorthandProperties;
+
+            return shorthandBuckets[propA] - shorthandBuckets[propB];
+          }
+          return 0;
+        });
+
+        const sourceCode = context.getSourceCode();
+        const sortedCode = sortedProperties
+          .map((property) => sourceCode.getText(property))
+          .join(', ');
+
+        // Replace the old object expression with the new sorted one
+        const newObjectExpression = `{ ${sortedCode} }`;
+
+        return fixer.replaceText(node, newObjectExpression);
+      }
+      return null;
+    },
+  });
+};
+
+const arePropertiesInTheRightOrder = (
+  node: ObjectExpression,
+  propertyA: Property,
+  i: number
+): boolean => {
+  if (node.properties.length > 0) {
+    if (propertyA.key.type === 'Identifier') {
+      const propA = kebabCase(propertyA.key.name) as ShorthandProperties;
+      const depthA = shorthandBuckets[propA];
+
+      for (let j = i + 1; j < node.properties.length; j++) {
+        const propertyB = node.properties[j];
+
+        if (propertyB.type === 'Property') {
+          if (propertyB.key.type === 'Identifier') {
+            const propB = kebabCase(propertyB.key.name) as ShorthandProperties;
+            const depthB = shorthandBuckets[propB];
+
+            if (propA === propB) {
+              node.properties.splice(node.properties.indexOf(propertyB), 1);
+              continue;
+            }
+
+            const shorthandForResA = shorthandFor[propA];
+            const shorthandForResB = shorthandFor[propB];
+
+            if (
+              shorthandForResA !== true &&
+              shorthandForResA !== undefined &&
+              shorthandForResB !== true &&
+              shorthandForResB !== undefined
+            ) {
+              // find intersection between objects
+              const intersectionAB = shorthandForResA.filter((x) => shorthandForResB.includes(x));
```

Comment:
i wonder if converting both `shorthandForResA` and `shorthandForResB` to sets, and [then running `shorthandForResA.intersection(shorthandForResB)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/intersection) would be faster?

Reply:
I tried this initially and couldn't find the `intersection()` method on the `Set` type? 🤔

Reply:
ah available in node v22 https://caniuse.com/mdn-javascript_builtins_set_intersection , we are still on node v18 😓